### PR TITLE
Add connection properties for specifying custom TrustManager

### PR DESCRIPTION
**Pull request by sehrope**
This PR adds two new connection properties:

* `trustManagerClass` - The fully qualified class name of a custom javax.net.ssl.TrustManager
* `trustManagerConstructorArg` - An optional argument to pass to the constructor of the TrustManager.
If trustManagerClass is specified and an encrypted connection is requested, the custom TrustManager will be used rather than the default system JVM keystore based TrustManager.

The primary use case for this is to dynamically specify the specific set of certificates that are trusted on a per connection basis without modifying the global settings for the JVM environment (i.e. no need to use a shared / global JKS).

For the error message localization code on line 1803, I've added handling for the case where the caught exception does not provide a localized message. That situation can arise when a custom TrustManager implementation is used as the Driver has no control over that code (it's use specified).

The new code tries to get a localized message, if one is not available then it tries the non-localized message, and if that is not available it defaults to an empty string. The final empty string default is required as subsequent code assumes that all caught exceptions will have a localized error message and would otherwise throw a NullPointerException from within the error handling code.

There's no additional tests for it in the test suite but I've tested it externally using both the specific custom TrustManager we'll be using as well as custom implementation of PermissiveX509TrustManager (which effectively mimics the behavior of trustServerCertificate=true).

See the prior PR submitted to the master branch, #70, for additional commentary and an example use case.

### CODE REVIEW

**Comment by sehrope** [comment]
I've rebased this atop the formatted dev branch so it applies cleanly.

Any idea when / if this is going to get merged? It's been in the "*Under review*" tag for two months.

**Comment by ajlam** [comment]
@sehrope - we are currently considering the PR and trying to understand if there are potential security implications of providing support for third party trust managers.

**Comment by sehrope** [comment]
Why was this closed? I'd still like to get something that provides this type of extensibility added to the driver At the moment it's not possible as you'd have to muck with the local JVM wide keystore.

**Comment by v-xiangs** [comment]
Sorry @sehrope , I missed your last comment. We removed our previous Dev branch and all PRs created for Dev were closed automatically. We have re-opened the PR. I am sorry for the trouble.

We are looking at this PR, we will get back to you soon. Thank you.

**Comment by sehrope** [comment]
Great. It's been a while since this was put together. I think it still applies cleanly but if not let me know and I'll rebase it.

**Comment by ulvii** [comment]
Hi @sehrope ,
I have been doing some internal testing and your changes look good so far. 
Please also make sure that, you add the new connection properties to the `SQLServerDataSource` too. You can have a look at #422 for guidance.

File: src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java [hunk 1]
```diff
@@ -1618,7 +1618,21 @@ void enableSSL(String host,
 
                 tm = new TrustManager[] {new PermissiveX509TrustManager(this)};
             }
-
+            // Otherwise, we'll check if a specific TrustManager implemenation has been requested and
+            // if so instantiate it, optionally specifying a constructor argument to customize it.
+            else if (con.getTrustManagerClass() != null)
+            {
+                Class<?> tmClazz = Class.forName(con.getTrustManagerClass());
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java]: Please rename `tmClazz` to `tmClass` :)

**Comment by ulvii** [comment]
@sehrope,
I created a [PR](https://github.com/sehrope/mssql-jdbc/pull/1) to your repository to add a few basic verification tests. You should be able to merge it to custom-trust-manager branch. It would also be great if you could add more test cases.

**Comment by sehrope** [comment]
Yes I saw and submitted some feedback on it. Overall looks great!

I'll get that merged, update the naming of clazz/class, and then push a new version. Should have it out some time tomorrow.

File: src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java [hunk 2]
```diff
@@ -1618,7 +1618,21 @@ void enableSSL(String host,
...
+                if (!TrustManager.class.isAssignableFrom(tmClazz)) {
+                    throw new IllegalArgumentException("The class specified by the trustManagerClass property must implement javax.net.ssl.TrustManager");
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java]: Please also use the Microsoft code [formatter ](https://github.com/Microsoft/mssql-jdbc/blob/master/mssql-jdbc_formatter.xml)for the changes.

**Comment by ulvii** [comment]
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla.microsoft.com.

**Comment by sehrope** [comment]
Pushed a bunch of formatting and test cleanup related updates.

Ideally I'd like to add a more complicated integration test that validates the server's specific certificate (or CA) but that will have to wait as I don't think the CI infrastructure can support that yet. It'd require changing the startup of the test server to either use an explicit cert or retrieve it prior to the test being run so both the Travis and AppVeyor config would be modified.

**Comment by sehrope** [comment]
> To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla.microsoft.com. Once you sign the agreement, the cla-required label will automatically be removed.

I signed the CLA a while back but I think this PR predates the setup of the bot so it's likely confused. You may need to manually remove the tag.

**Comment by ulvii** [comment]
Hi @sehrope,

Please reformat your changes to use Microsoft formatter. It can be found [here](https://github.com/Microsoft/mssql-jdbc/blob/master/mssql-jdbc_formatter.xml).

I did some more testing using actual certificates and your changes look good. The PR will be merged once we finalize testing. Great contribution 👍

**Comment by sehrope** [comment]
Formatting is complete. To minimize the diff I only ran the formatter on the changes themselves. Surrounding code was not reformatted.

Also, the enum declarations in SQLServerDriver don't completely align anymore because the name of the new one is longer than the rest (line 281). I'd suggest having that corrected in a separate PR as otherwise the diff would clobber any other PRs that are adding new properties.

**Comment by sehrope** [comment]
Awesome. Thanks for merging this in!

**Pull request merged**